### PR TITLE
fix: handle template literals in startsWith

### PR DESCRIPTION
**Pull request by jly36963**
Relates to: https://github.com/gustavoguichard/string-ts/issues/181

## Description

I needed to have more granular control over the comparison, so I replaced the 
`${S}${string}` 
part with 
`${infer THead}${infer TRest}`

I needed to remove `Length<T>` in a lot of places -- it doesn't work with template literals.

## Details

`Length` does not (and cannot) handle string interpolation well.
```ts
Length<'abc'> // 3
Length<`abc${string}`> // number
```
`Slice<T, startIndex, endIndex>` uses `Length<T>` as the default value for `endIndex`. 
Both `StartsWith`/`EndsWith` use `Slice`.

I refactored `Slice`/`StartsWith`/`EndsWith` to not use `Length` (or only use it when `T` is known to be a literal) and I now have them working with template literals.

Also, `slice`/`startsWith`/`endsWith` are not methods with default number arguments (eg: `0` or `Length<T>`), they are overloads where the last arguments are optionally provided.

Consider
```ts
slice(start?: number | undefined, end?: number | undefined): string
startsWith(searchString: string, position?: number | undefined): boolean
endsWith(searchString: string, endPosition?: number | undefined): boolean
```

In other words, updating the params for these types to be
`P: number | undefined = undefined`
instead of
`P: number = Length<T>`
makes this fix possible _and_ more closely aligns our types with the TS definitions for those string methods.

### CODE REVIEW

**Review by gustavoguichard — APPROVED** [review]
All good to me! The code is more complex than what I wanted but I trust you took the right decisions!
TY very much for the PR!!
